### DISCUSSION TITLE
Can't get wezterm to exit neoVim using CWD even though it recognizes it.

### DISCUSSION THREAD

**Discussion by allreddv** [wezterm/wezterm#6272] [Q&A] [answered]
### What Operating System(s) are you running on?

macOS

### WezTerm version

20240203-110809-5046fc22

### Ask your question!

I have spent way too much time trying to get this to work, so finally reaching out for help.  When i exit nvim I want the wezterm window nvim was in to be at the current working directory.  I have been successful with integrating wezterm and the shell (zsh) using the .wezterm_shell_integration.sh  .  I also set it up in nvim to send OSC 7 escape sequences to wezterm everytime CWD is changed within neoVim and this is working. This is confirmed both by showing the changed CWD in the status bar and i am also able to open new tabs/panes to that CWD even when changed in neoVim.

The issue I am having is that when i exit nvim, I would like the tab/pane that nvim was in to exit to the CWD. I would have thought i could do so using the "set-working-directory" command on nvim exit, but I can't get it to work.  So what happens is that on nvim close the prompt is at the original directory of when that wezterm window was opened even though that is no longer the CWD.  

Even if i CD to a new directory, it doesn't change the CWD which is reflected both in the status bar which shows the CWD from where i was in neoVim and also if I open a new pane, that pane will be in the CWD from  when neoVim exited, and if I CD to a new directory from that new pane then the CWD will change, again reflected both in the status bar and if I then open additional panes.  

I have even tried using barebones testing .zshrc and wezterm.lua files with just enough to have the shell integration and ability to receive the OSC 7 escape sequences just to see if i was having conflict issues with any other of my configurations, but still the same issue.  

Thaniks in advance for any help.  My ADHD hyper focus kicked in yesterday and i lost countless hours trying to ge this to work when i should have just been coding the project i was supposed to be working on.  hahah

**Comment by loops** [comment]
Hi allreddv,

Unfortunately, this is not directly possible.  Wezterm can not alter the execution environment of an already running process.

You would have to configure your shell to change the cwd itself, which is possible with some gymnastics, but depends on which shell you're using, and how much tomfoolery you're willing to tolerate ;-)

**Reply by allreddv** [reply]
Okay, thanks, at least I can stop trying the route I was trying.  I might give a shot seeing if I can have the shell do it just for fun.  Unless you know for sure already can't do that with zsh, then I won't even try.  The other thing I might try to do is open a new pane on exit of nvim and close pane nvim was running in because that new pane would be in the CWD, because that is already working, just have not tried the closing and opening of panes  on nvim exit.   

Amazing terminal by the way.  Absolutely love it. 

 Only current request I would like to see is ability to change cursor colors dynamically.  I would like to change the cursor color depending on vim mode I am in. I am able to change everything else in nvim based on the mode, but not cursor as that is coming directly from wezterm.  Not a big deal, but we can always wish.

**Reply by loops** [reply]
It's possible with zsh.   If you define a function named precmd it will run just before your prompt is displayed.  In it you can write code to discover the directory nvim was in at exit, and cd to it.   This implies you need to get nvim to write out its cwd somehow; either to a file, or if you want to get fancy, into a wezterm user variable.        Good luck :)

There's an easier answer for your Neovim cursor color request.   Look in Neovim's     :help faq
under the heading "HOW TO CHANGE CURSOR COLOR IN THE TERMINAL".

**Reply by loops** [reply]
Hey allreddv,

This kept sneaking into my thoughts, so I tried out the precmd method of zsh.

Changing .zshrc for Zsh, so that just before each prompt it will cd to whatever gets put into the SAVECWD file; and then
delete SAVECWD:

    export SAVECWD=~/.savecwd.$$
    function precmd () {
          [[ -f "$SAVECWD" ]] && cd "$(<"$SAVECWD")" && noglob unlink -- "$SAVECWD"
    }

And then adding this to Neovim's init.lua, to save its current directory if the
environment variable SAVECWD is set:

    vim.api.nvim_create_autocmd("VimLeavePre", {                
      callback = function()                                     
        local save_cwd = vim.fn.getenv("SAVECWD")               
        if save_cwd and save_cwd ~= "" then                     
          local cwd = vim.fn.getcwd()                           
          local file = io.open(save_cwd, "w")                   
          if file then                                          
            file:write(cwd)                                     
            file:close()                                        
          else                                                  
            print("SAVECWD: Unable to open file: " .. save_cwd) 
          end                                                   
        end                                                     
      end,                                                      
    })                                                          


Seems to do what you want.   But it's not the most secure thing ever, and is not something that should be put on a production machine anywhere...

HTH

**Reply by allreddv** [reply]
Thanks so much, i will take a look at it tomorrow as i can't at the moment.  But, also about the cursor color.  I had looked at the docs before but couldn't get to work. I had the shape changing but not the color.   But after you said it could be done, I took a look again and got it to work.  So thanks for that.  I will let you know tomorrow about the CWD bit.  Thanks again, really appreciate it.

**Reply by allreddv** [reply]
I tested out my idea of opening a pane and then closing the pane that nvim was running in.  I just have a simple script

   1   │ #!/bin/bash
   2   │
   3   │ # Run Neovim 
   4   │ nvim "$@"
   5   │
   6   │ # After Neovim exits, open a new pane in WezTerm with Zsh
   7   │ wezterm cli split-pane --horizontal -- zsh
   8   │
   9   │ # Exit the current pane (terminate current shell)
  10   │ wezterm cli kill-pane
  
  
  and then created an alias for the nvim_wezterm.sh script and i just run that to open nvim.  Seems to be working perfectly because you can't even notice it since it is doing a quick split and then closes the one nvim was in,  so in the end it is exactly like it was, even if i have a bunch of other panes open it doesn't affect them. 
  
  Do you see anything wrong with the way I am doing this? I am fairly new to working in the terminal and nvim so not sure if this is a bad idea for some reason to wrap nvim in a shell script.

**Reply by loops** [reply, marked as answer]
If it works for you, that's all that matters!   One thing to check, is if it disrupts your command history.  That'll depend on how you have zsh configured, and what you're used to.

**Reply by allreddv** [reply]
Cool, going with this option since is simple and working.  Thanks for all your help, especially with the colored cursor.  It was annoying me because i change the status bar color, line number color, and shape of the cursor but the color remained the same.  All working now.
